File: backend/go/internal/middleware/tracing.go
package middleware

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	sdkresource "go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
	"go.opentelemetry.io/otel/trace"
)

// TracingConfig configures the OpenTelemetry tracing provider.
type TracingConfig struct {
	ServiceName string
	Endpoint    string
	Insecure    bool
	SampleRate  float64
}

// InitTracer creates and registers a global OpenTelemetry tracer provider.
// Returns a shutdown function that must be called on application exit.
func InitTracer(cfg TracingConfig) (func(context.Context) error, error) {
	ctx := context.Background()

	var opts []otlptracegrpc.Option
	if cfg.Endpoint != "" {
		opts = append(opts, otlptracegrpc.WithEndpoint(cfg.Endpoint))
	}
	if cfg.Insecure {
		opts = append(opts, otlptracegrpc.WithInsecure())
	}

	exporter, err := otlptracegrpc.New(ctx, opts...)
	if err != nil {
		return nil, fmt.Errorf("create OTel trace exporter: %w", err)
	}

	res, err := sdkresource.New(ctx,
		sdkresource.WithAttributes(
			semconv.ServiceNameKey.String(cfg.ServiceName),
			semconv.ServiceVersionKey.String("1.0.0"),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("create OTel resource: %w", err)
	}

	sampler := sdktrace.AlwaysSample()
	if cfg.SampleRate > 0 && cfg.SampleRate < 1 {
		sampler = sdktrace.TraceIDRatioBased(cfg.SampleRate)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sampler),
	)

	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	return tp.Shutdown, nil
}

// TracingMiddleware creates a chi-compatible middleware that starts an OTel span
// for each HTTP request. It records method, URL, host, user agent, and status code.
func TracingMiddleware(serviceName string) func(http.Handler) http.Handler {
	tracer := otel.Tracer(serviceName)

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ctx, span := tracer.Start(
				r.Context(),
				r.Method+" "+routePattern(r),
				trace.WithAttributes(
					attribute.String("http.method", r.Method),
					attribute.String("http.url", r.URL.String()),
					attribute.String("http.host", r.Host),
					attribute.String("http.user_agent", r.UserAgent()),
				),
			)
			defer span.End()

			ww := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}
			next.ServeHTTP(ww, r.WithContext(ctx))

			span.SetAttributes(attribute.Int("http.status_code", ww.statusCode))
			if ww.statusCode >= 500 {
				span.SetAttributes(attribute.Bool("error", true))
			}
		})
	}
}

// SpanFromContext returns the active span from the context.
func SpanFromContext(ctx context.Context) trace.Span {
	return trace.SpanFromContext(ctx)
}

// TraceIDFromContext extracts the trace ID as a hex string.
func TraceIDFromContext(ctx context.Context) string {
	span := trace.SpanFromContext(ctx)
	if sc := span.SpanContext(); sc.HasTraceID() {
		return sc.TraceID().String()
	}
	return ""
}

// SpanIDFromContext extracts the span ID as a hex string.
func SpanIDFromContext(ctx context.Context) string {
	span := trace.SpanFromContext(ctx)
	if sc := span.SpanContext(); sc.HasSpanID() {
		return sc.SpanID().String()
	}
	return ""
}

// StartSpan creates a child span from the given context.
func StartSpan(ctx context.Context, name string, attrs ...attribute.KeyValue) (context.Context, trace.Span) {
	return otel.Tracer("sentinel").Start(ctx, name, trace.WithAttributes(attrs...))
}

// MeasureDBCall instruments a database call with tracing and metrics.
func MeasureDBCall(ctx context.Context, operation string, fn func(context.Context) error) error {
	ctx, span := StartSpan(ctx, "db."+operation,
		attribute.String("db.operation", operation),
		attribute.String("db.system", "dynamodb"),
	)
	defer span.End()

	start := time.Now()
	err := fn(ctx)
	RecordDBQuery(operation, time.Since(start), err)

	if err != nil {
		span.RecordError(err)
	}
	return err
}

File: backend/python/analytics_service/tracing.py
"""OpenTelemetry distributed tracing for the Sentinel Analytics Service."""

import os
from typing import Optional

from fastapi import FastAPI
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.resources import Resource, SERVICE_NAME
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor


def setup_tracing(
    app: FastAPI,
    endpoint: Optional[str] = None,
    service_name: str = "sentinel-analytics",
) -> Optional[callable]:
    """
    Configure OpenTelemetry tracing for a FastAPI application.

    Args:
        app: The FastAPI application instance.
        endpoint: OTel Collector gRPC endpoint (e.g. "otel-collector:4317").
                  If None, tracing is disabled.
        service_name: Service name for span attributes.

    Returns:
        A shutdown callable, or None if tracing is disabled.
    """
    if endpoint is None:
        endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
        if not endpoint or endpoint == "none":
            return None

    resource = Resource.create({SERVICE_NAME: service_name})

    exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
    processor = BatchSpanProcessor(exporter)

    provider = TracerProvider(resource=resource)
    provider.add_span_processor(processor)
    trace.set_tracer_provider(provider)

    FastAPIInstrumentor.instrument_app(app)

    def shutdown():
        provider.shutdown()

    return shutdown


def get_trace_id() -> str:
    """Extract the current trace ID from the active span context."""
    span = trace.get_current_span()
    ctx = span.get_span_context()
    if ctx.is_valid:
        return format(ctx.trace_id, "032x")
    return ""
